### ISSUE TITLE
protobuf test DiskSourceTreeTest.MapDirectory fails in specific cases

### ISSUE THREAD

**Issue by Kissaki** [mumble-voip/mumble-releng#44]
The protobuf test DiskSourceTreeTest.MapDirectory fails when I run it from my cygwin shell, but succeeds when I run it from a fresh cmd or from a command prompt spawn with Cygwin.bat.

I am using a cygwin install installed and updated by [babun](http://babun.github.io/) (so I don't have to handle the GUI stuff but was able to initially install the required packages via command line). As it is a normal cygwin install which makes use of the cygwin setup.exe this shouldn't make a difference as far as the cygwin install goes.

I updated to protobuf from 2.6.1 to 3.2.0, and the test still fails.

~~I also noticed the issue #5 still exists and fails, unless the patch is applied (for both versions of protobuf). However, I have not verified this is linked in any way.~~

> cmd /c tests.exe --gtest_filter=DiskSourceTreeTest.MapDirectory

**Comment by Kissaki** [comment]
```
{ win32-static } master » cmd /c "cd ../../../../../win64-static-no-ltcg-1.3.x-2017-03-04-1ddd966-811.build\protobuf-3.2.0\cmake && tests.exe --gtest_filter=DiskSourceTreeTest.MapDirectory"
Running main() from gmock_main.cc
Note: Google Test filter = DiskSourceTreeTest.MapDirectory
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DiskSourceTreeTest
[ RUN      ] DiskSourceTreeTest.MapDirectory
C:\MumbleBuild\win64-static-no-ltcg-1.3.x-2017-03-04-1ddd966-811.build\protobuf-3.2.0\src\google\protobuf\compiler\importer_unittest.cc(289): error: Value of: input == NULL
  Actual: false
Expected: true
C:\MumbleBuild\win64-static-no-ltcg-1.3.x-2017-03-04-1ddd966-811.build\protobuf-3.2.0\src\google\protobuf\compiler\importer_unittest.cc(290): error: Value of: source_tree_.GetLastErrorMessage()
  Actual: "Backslashes, consecutive slashes, \".\", or \"..\" are not allowed in the virtual path"
Expected: error_message
Which is: "File not found."
[  FAILED  ] DiskSourceTreeTest.MapDirectory (15 ms)
[----------] 1 test from DiskSourceTreeTest (15 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (16 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DiskSourceTreeTest.MapDirectory

 1 FAILED TEST
```

**Comment by Kissaki** [comment]
It occurs on the [last line/test of the test](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer_unittest.cc#L331):

> ExpectCannotOpenFile("baz/foo/", "File not found.");

**Comment by Kissaki** [comment]
* [TEST_F(DiskSourceTreeTest, MapDirectory)](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer_unittest.cc#L331)
* [ExpectCannotOpenFile](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer_unittest.cc#L286)
* [DiskSourceTree::Open](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer.cc#L434)
* [DiskSourceTree::OpenVirtualFile](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer.cc#L442)
* [CanonicalizePath](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/compiler/importer.cc#L275)
* [Split](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/stubs/strutil.h#L234)
* [SplitStringUsing](https://github.com/google/protobuf/blob/39f9b43219bc5718b659ed72a2130a7b2ce66108/src/google/protobuf/stubs/strutil.cc#L194)

Expecting the canonalized path to not have the slash at the end and thus be different from the passed path, thus the error should be thrown.

Does cygwin in one instance convert it to backslashes?

**Comment by mkrautz** [comment]
Fixed in the tree via 97235e334fee00ae2eef6fd66fa3feeb5eef9562
